src/libopensc/padding.c

```diff
@@ -231,22 +237,200 @@ int sc_pkcs1_strip_digest_info_prefix(unsigned int *algorithm,
+	if (2 * (unsigned)hlen + 2 > mod_length)
+		/* RSA key too small for chosen hash (1296 bits or higher needed for
+		 * signing SHA-512 hashes) */
+		return SC_ERROR_NOT_SUPPORTED;
+
+	if (RAND_bytes(salt, hlen) != 1)
+		return SC_ERROR_INTERNAL;
+
+	/* Hash M' to create H */
+	if (!(ctx = EVP_MD_CTX_create()))
+		goto done;
+	EVP_DigestInit_ex(ctx, md, NULL);
+	memset(buf, 0x00, 8);
+	EVP_DigestUpdate(ctx, buf, 8);
+	EVP_DigestUpdate(ctx, in, hlen); /* mHash */
+	EVP_DigestUpdate(ctx, salt, hlen);
+
+	/* Construct padding2, salt, H, and BC in the output block */
+	/* DB = PS || 0x01 || salt */
+	memset(out, 0x00, plen - 1); /* emLen - sLen - hLen - 2 */
+	out[plen - 1] = 0x01;
+	memcpy(out + plen, salt, hlen);
+	EVP_DigestFinal_ex(ctx, out + dblen, NULL); /* H */
+	out[dblen + hlen] = 0xBC;
+	/* EM = DB* || H || 0xbc
+	 *  *the first part is masked later */
+
+	/* Construct the DB mask block by block and XOR it in. */
+	mgf1_md = mgf1_flag2md(mgf1_hash);
+	if (mgf1_md == NULL)
+		return SC_ERROR_NOT_SUPPORTED;
+	mgf1_hlen = EVP_MD_size(mgf1_md);
+
+	mgf_rounds = (dblen + mgf1_hlen - 1) / mgf1_hlen; /* round up */
+	for (round = 0; round < mgf_rounds; ++round) {
+		buf[0] = (round&0xFF000000U) >> 24;
+		buf[1] = (round&0x00FF0000U) >> 16;
+		buf[2] = (round&0x0000FF00U) >> 8;
+		buf[3] = (round&0x000000FFU);
+		EVP_DigestInit_ex(ctx, mgf1_md, NULL);
+		EVP_DigestUpdate(ctx, out + dblen, hlen); /* H (Z parameter of MGF1) */
+		EVP_DigestUpdate(ctx, buf, 4); /* C */
+		EVP_DigestFinal_ex(ctx, mask, NULL);
+		/* this is no longer part of the MGF1, but actually
+		 * XORing mask with DB to create maskedDB inplace */
+		for (i = round * mgf1_hlen, j = 0; i < dblen && j < mgf1_hlen; ++i, ++j) {
+			out[i] ^= mask[j];
+		}
+	}
+
+	/* Set leftmost N bits in leftmost octet in maskedDB to zero
+	 * to make sure the result is smaller than the modulus ( +1)
+	 */
+	out[0] &= (0xff >> (8 * mod_length - mod_bits + 1));
+
+	*out_len = mod_length;
+	rv = SC_SUCCESS;
+
+done:
+	OPENSSL_cleanse(salt, sizeof(salt));
+	OPENSSL_cleanse(mask, sizeof(mask));
+	if (ctx) {
+		EVP_MD_CTX_destroy(ctx);
+	}
+	return rv;
+}
+
+static int hash_len2algo(size_t hash_len)
+{
+	switch (hash_len) {
+	case SHA_DIGEST_LENGTH:
+		return SC_ALGORITHM_RSA_HASH_SHA1;
+	case SHA224_DIGEST_LENGTH:
+		return SC_ALGORITHM_RSA_HASH_SHA224;
+	case SHA256_DIGEST_LENGTH:
+		return SC_ALGORITHM_RSA_HASH_SHA256;
+	case SHA384_DIGEST_LENGTH:
+		return SC_ALGORITHM_RSA_HASH_SHA384;
+	case SHA512_DIGEST_LENGTH:
+		return SC_ALGORITHM_RSA_HASH_SHA512;
+	}
+	/* Should never happen -- the mechanism and data should be already
+	 * verified to match one of the above. If not, we will fail later
+	 */
+	return SC_ALGORITHM_RSA_HASH_NONE;
+}
+#endif
+
 /* general PKCS#1 encoding function */
 int sc_pkcs1_encode(sc_context_t *ctx, unsigned long flags,
-	const u8 *in, size_t in_len, u8 *out, size_t *out_len, size_t mod_len)
+	const u8 *in, size_t in_len, u8 *out, size_t *out_len, size_t mod_bits)
```

Comment:
True. Fixed in other cases, but I am not sure what to do with the minidriver -- it is passing the last argument of 0, which is not valid argument in `sc_pkcs1_add_01_padding()`. Is this code path ever executed?

Reply:
The minidriver only sets hash flags, no padding flags. Hence, the `mod_len` argument is not used in `sc_pkcs1_encode()`/`sc_pkcs1_add_01_padding()`. So regarding your PR, you could ignore the use of `sc_pkcs1_encode()` in minidriver.c.

However, I'm not sure if the minidriver implementation is optimal. Because after calling `sc_pkcs1_encode()` with the flag for hashing:
https://github.com/OpenSC/OpenSC/blob/3cc2670f3e5f30d0d244b2d07d9f959f3e840af7/src/minidriver/minidriver.c#L4742-L4762
an additional call to sc_pkcs15_compute_signature is made
https://github.com/OpenSC/OpenSC/blob/3cc2670f3e5f30d0d244b2d07d9f959f3e840af7/src/minidriver/minidriver.c#L4823-L4830
the second time, only the padding flag is used. I think the same could have been achieved by just calling `sc_pkcs15_compute_signature()` with both, padding and hashing flags. `sc_pkcs15_compute_signature()` will then call `sc_pkcs1_encode()` if needed. @vletoux, what do you think?